### PR TITLE
API changes to sync with one Participant per Context change in rmw_fastrtps

### PR DESCRIPTION
**Pull request by ivanpauno**
Needed since https://github.com/ros2/rmw/pull/189.

### CODE REVIEW

**Comment by eboasson** [comment]
It certainly looks perfectly reasonable to me. The answer to @hidmic's question is a bit more difficult: this involves a few rather large pull requests and we need to first figure out what changes are actually needed. The change to using a single participant is trivial, it is the bit that relates to introspection that is less obvious.

Currently, the Cyclone RMW layer just queries its discovery data on demand and otherwise ignores all changes save for triggering the graph guard condition. I guess it will now have to call  functions in ``rmw_dds_common`` for each change, shuttle the new discovery topics back-and-forth, and forward the introspection operations to it. How much work those changes will be I don't know without having a good look at the interfaces.

So the only thing I can say definitively is that it we'll be looking into the details and hope to be able to this soon.

**Comment by ivanpauno** [comment]
> The change to using a single participant is trivial, it is the bit that relates to introspection that is less obvious.

Yes, completly agreed.

> How much work those changes will be I don't know without having a good look at the interfaces.

Please, let me know if I can help, or if anything has to be modified in `rmw_dds_common` to make it more suitable for `cyclonedds`.

> So the only thing I can say definitively is that it we'll be looking into the details and hope to be able to this soon.

Great, thanks!

File: rmw_cyclonedds_cpp/src/rmw_node.cpp [hunk 1]
```diff
@@ -2932,7 +2965,7 @@ extern "C" rmw_ret_t rmw_get_node_names(
   }
 
   std::vector<std::pair<std::string, std::string>> ns;
-  const auto re = std::regex("^name=(.*);namespace=(.*);$", std::regex::extended);
+  const auto re = std::regex("^name=([^;]*);namespace=([^;]*);", std::regex::extended);
```
Comment (hidmic) [inline rmw_cyclonedds_cpp/src/rmw_node.cpp]: @ivanpauno a hidden bugfix? Maybe we can keep this one in its own commit and rebase-merge afterwards?

File: rmw_cyclonedds_cpp/src/rmw_node.cpp [hunk 2]
```diff
@@ -2987,6 +3020,87 @@ extern "C" rmw_ret_t rmw_get_node_names(
   return RMW_RET_BAD_ALLOC;
 }
 
+extern "C" rmw_ret_t rmw_get_node_names_with_security_contexts(
```
Comment (hidmic) [inline rmw_cyclonedds_cpp/src/rmw_node.cpp]: @ivanpauno should this (and every other change in this patch) be conditionally compiled based on RMW version too?

File: rmw_cyclonedds_cpp/src/rmw_node.cpp [hunk 3]
```diff
@@ -2987,6 +3020,87 @@ extern "C" rmw_ret_t rmw_get_node_names(
...
+  auto node_impl = static_cast<CddsNode *>(node->data);
+  if (rmw_check_zero_rmw_string_array(node_names) != RMW_RET_OK ||
+    rmw_check_zero_rmw_string_array(node_namespaces) != RMW_RET_OK)
+  {
+    return RMW_RET_ERROR;
+  }
+
+  std::vector<std::tuple<std::string, std::string, std::string>> ns;
+  const auto re = std::regex(
+    "^name=([^;]*);namespace=([^;]*);securitycontext=([^;]*);", std::regex::extended);
```
Comment (hidmic) [inline rmw_cyclonedds_cpp/src/rmw_node.cpp]: @ivanpauno nit: why not calling the constructor directly i.e. `std::regex re(...);`?

File: rmw_cyclonedds_cpp/src/rmw_node.cpp [hunk 4]
```diff
@@ -2987,6 +3020,87 @@ extern "C" rmw_ret_t rmw_get_node_names(
...
+  rcutils_allocator_t allocator = rcutils_get_default_allocator();
+  if (rcutils_string_array_init(node_names, ns.size(), &allocator) != RCUTILS_RET_OK ||
+    rcutils_string_array_init(node_namespaces, ns.size(), &allocator) != RCUTILS_RET_OK ||
+    rcutils_string_array_init(security_contexts, ns.size(), &allocator) != RCUTILS_RET_OK)
+  {
+    RMW_SET_ERROR_MSG(rcutils_get_error_string().str);
+    goto fail_alloc;
+  }
+  size_t i;
+  i = 0;
```
Comment (hidmic) [inline rmw_cyclonedds_cpp/src/rmw_node.cpp]: @ivanpauno nit: last two lines can be collapsed into one.

File: rmw_cyclonedds_cpp/src/rmw_node.cpp [hunk 5]
```diff
@@ -2987,6 +3020,87 @@ extern "C" rmw_ret_t rmw_get_node_names(
...
+extern "C" rmw_ret_t rmw_get_node_names_with_security_contexts(
+  const rmw_node_t * node,
+  rcutils_string_array_t * node_names,
+  rcutils_string_array_t * node_namespaces,
+  rcutils_string_array_t * security_contexts)
+{
+  RET_WRONG_IMPLID(node);
+  auto node_impl = static_cast<CddsNode *>(node->data);
+  if (rmw_check_zero_rmw_string_array(node_names) != RMW_RET_OK ||
+    rmw_check_zero_rmw_string_array(node_namespaces) != RMW_RET_OK)
```
Comment (hidmic) [inline rmw_cyclonedds_cpp/src/rmw_node.cpp]: @ivanpauno should the same check be performed on `security_contexts` if not `nullptr`?

File: rmw_cyclonedds_cpp/src/rmw_node.cpp [hunk 6]
```diff
@@ -2987,6 +3020,87 @@ extern "C" rmw_ret_t rmw_get_node_names(
...
+  if (rcutils_string_array_init(node_names, ns.size(), &allocator) != RCUTILS_RET_OK ||
+    rcutils_string_array_init(node_namespaces, ns.size(), &allocator) != RCUTILS_RET_OK ||
+    rcutils_string_array_init(security_contexts, ns.size(), &allocator) != RCUTILS_RET_OK)
+  {
+    RMW_SET_ERROR_MSG(rcutils_get_error_string().str);
+    goto fail_alloc;
+  }
+  size_t i;
+  i = 0;
+  for (auto && n : ns) {
```
Comment (hidmic) [inline rmw_cyclonedds_cpp/src/rmw_node.cpp]: @ivanpauno why rvalue references?

File: rmw_cyclonedds_cpp/src/rmw_node.cpp (hunk 1 above)
Comment (ivanpauno) [reply rmw_cyclonedds_cpp/src/rmw_node.cpp]: Not exactly, as it wasn't a bug.
I'm adding `securitycontext=CONTEXT_NAME;` at the end of user data.
The changes are needed because of:
- I want the regex to match `name` and `namespace` also in old versions that doesn't have the `securitycontext` part at the end. So I dropped the `$` from the regex.
- To match correctly with `name=NODE_NAME;namespace=NAMESPACE;securitycontext=CONTEXT;`, I needed to replace the `(.*)` groups with `([^;]*)`. If not, the result would be `name="NODE_NAME"`, `namespace="NAMESPACE;securitycontext=CONTEXT"`.

File: rmw_cyclonedds_cpp/src/rmw_node.cpp (hunk 2 above)
Comment (ivanpauno) [reply rmw_cyclonedds_cpp/src/rmw_node.cpp]: Thanks for the reminder, see https://github.com/ros2/rmw_cyclonedds/pull/106/commits/756fa27bc547a00ad36b80327ef825185a205d87.

File: rmw_cyclonedds_cpp/src/rmw_node.cpp (hunk 3 above)
Comment (ivanpauno) [reply rmw_cyclonedds_cpp/src/rmw_node.cpp]: preexisting, I can address it if you mind strongly.

P.S.: As file as I understand, this is calling just the constructor.

File: rmw_cyclonedds_cpp/src/rmw_node.cpp (hunk 4 above)
Comment (ivanpauno) [reply rmw_cyclonedds_cpp/src/rmw_node.cpp]: Same as above, it comes from `rmw_get_node_names`.

File: rmw_cyclonedds_cpp/src/rmw_node.cpp (hunk 5 above)
Comment (ivanpauno) [reply rmw_cyclonedds_cpp/src/rmw_node.cpp]: It seems to me that you're looking to the wrong function:
https://github.com/ros2/rmw_cyclonedds/blob/ad4879663c2244ff88604292b239cc0400bf94df/rmw_cyclonedds_cpp/src/rmw_node.cpp#L3093-L3100

File: rmw_cyclonedds_cpp/src/rmw_node.cpp (hunk 6 above)
Comment (ivanpauno) [reply rmw_cyclonedds_cpp/src/rmw_node.cpp]: Copied from `rmw_get_node_names`.

P.S.: I can maybe do something better than copy-pasting ...

File: rmw_cyclonedds_cpp/src/rmw_node.cpp (hunk 3 above)
Comment (hidmic) [reply rmw_cyclonedds_cpp/src/rmw_node.cpp]: Yeah, copy elision will eventually do the job but it's odd. OK to leave as-is if pre-existing.

File: rmw_cyclonedds_cpp/src/rmw_node.cpp (hunk 5 above)
Comment (hidmic) [reply rmw_cyclonedds_cpp/src/rmw_node.cpp]: I think not. The snippet you refer to takes care of finalizing it if present, but I don't see a `rmw_check_zero_rmw_string_array` counterpart and [right below](https://github.com/ros2/rmw_cyclonedds/blob/ad4879663c2244ff88604292b239cc0400bf94df/rmw_cyclonedds_cpp/src/rmw_node.cpp#L3057) what is referred to by `security_context` is initialized without checking whether it was `nullptr` or not.

Comment (ivanpauno) [reply rmw_cyclonedds_cpp/src/rmw_node.cpp]: See https://github.com/ros2/rmw_cyclonedds/pull/106/commits/1c072f236b94853a1042b468215625726d5c8011.

File: rmw_cyclonedds_cpp/src/rmw_node.cpp (hunk 3 above)
Comment (eboasson) [reply rmw_cyclonedds_cpp/src/rmw_node.cpp]: Don't feel like you have to leave non-idiomatic C++ bits untouched. Chances are that I wrote them, and as I'm more used to C than to C++ and have grown to dislike both, my C++ tends to come a bit weird sometimes.

File: rmw_cyclonedds_cpp/src/rmw_node.cpp (hunk 6 above)
Comment (ivanpauno) [reply rmw_cyclonedds_cpp/src/rmw_node.cpp]: See https://github.com/ros2/rmw_cyclonedds/pull/106/commits/3c2f89c4414f5f9df88b3bc7d71197b0a069152d.

File: rmw_cyclonedds_cpp/src/rmw_node.cpp (hunk 4 above)
Comment (ivanpauno) [reply rmw_cyclonedds_cpp/src/rmw_node.cpp]: There is a pretty fun reason: when done in the same line, the compiler fails saying that the above `goto` is skipping an initialization. In general, it's safe to skip the initialization of a trivially destructible object (AFAIU), but the compiler shows the warning anyways (and we transform that warning in an error, because of our compiler flags).

When split in two lines, the compiler doesn't show a warning.

P.S.: This kind of warning never happens in pure C code, it only applies when you use C++.

File: rmw_cyclonedds_cpp/src/rmw_node.cpp (hunk 3 above)
Comment (ivanpauno) [reply rmw_cyclonedds_cpp/src/rmw_node.cpp]: Done in https://github.com/ros2/rmw_cyclonedds/pull/106/commits/3c2f89c4414f5f9df88b3bc7d71197b0a069152d.

File: rmw_cyclonedds_cpp/src/rmw_node.cpp (hunk 4 above)
Comment (hidmic) [reply rmw_cyclonedds_cpp/src/rmw_node.cpp]: Hmm interesting, odd but interesting.

Comment (wjwwood) [reply rmw_cyclonedds_cpp/src/rmw_node.cpp]: Please define ‘i’ higher in the scope to avoid this instead. This what is done in all the other functions that use goto.

File: rmw_cyclonedds_cpp/src/rmw_node.cpp [hunk 7]
```diff
@@ -3188,6 +3127,37 @@ extern "C" rmw_ret_t rmw_get_node_names_with_security_contexts(
...
+#if RMW_VERSION_GTE(0, 8, 2)
+extern "C" rmw_ret_t rmw_get_node_names_with_security_contexts(
+  const rmw_node_t * node,
+  rcutils_string_array_t * node_names,
+  rcutils_string_array_t * node_namespaces,
+  rcutils_string_array_t * security_contexts)
+{
+  if (RMW_RET_OK != rmw_check_zero_rmw_string_array(security_contexts)) {
+    return RMW_RET_ERROR;
+  }
```
Comment (hidmic) [inline rmw_cyclonedds_cpp/src/rmw_node.cpp]: @ivanpauno is it invalid to call `rmw_get_node_names_with_security_contexts()` with a null `security_contexts`? If not, I think this check should be relocated to the implementation function and conditionally executed based on `security_contexts` value.

File: rmw_cyclonedds_cpp/src/rmw_node.cpp [hunk 8]
```diff
@@ -3037,27 +3037,28 @@ static rmw_ret_t do_for_node_user_data(
...
-  std::vector<std::pair<std::string, std::string>> ns;
-  const auto re = std::regex("^name=([^;]*);namespace=([^;]*);", std::regex::extended);
+  std::vector<std::tuple<std::string, std::string, std::string>> ns;
   auto oper =
     [&ns, re](const dds_builtintopic_participant_t & sample, const char * ud) -> bool {
       std::cmatch cm;
       static_cast<void>(sample);
       if (std::regex_search(ud, cm, re)) {
-        ns.push_back(std::make_pair(std::string(cm[1]), std::string(cm[2])));
+        ns.push_back(std::make_tuple(std::string(cm[1]), std::string(cm[2]), std::string(cm[3])));
```
Comment (hidmic) [inline rmw_cyclonedds_cpp/src/rmw_node.cpp]: @ivanpauno this bit is still assuming that there's a third group, regardless of `security_contexts` value.

Is having a `security_context` stanza in `ud` mandatory? If not, why not simply making it optional in the regex? Then, `rmw_get_node_names(...)` reduces to `rmw_get_node_names_with_security_contexts(..., nullptr)`.

File: rmw_cyclonedds_cpp/src/rmw_node.cpp (hunk 7 above)
Comment (ivanpauno) [reply rmw_cyclonedds_cpp/src/rmw_node.cpp]: In `rmw_get_node_names_impl` `security_contexts` can be `NULL`.
In `rmw_get_node_names_with_security_contexts` `security_contexts` cannot be `NULL`.

I think the current logic is fine.

File: rmw_cyclonedds_cpp/src/rmw_node.cpp (hunk 8 above)
Comment (ivanpauno) [reply rmw_cyclonedds_cpp/src/rmw_node.cpp]: This is wrong, thanks for detecting it!

Comment (ivanpauno) [reply rmw_cyclonedds_cpp/src/rmw_node.cpp]: See https://github.com/ros2/rmw_cyclonedds/pull/106/commits/e09008b242ce747ee08ac0075ed8823d1253fa57.
Based on the [docs](https://en.cppreference.com/w/cpp/regex/match_results/operator_at), the previous code wasn't wrong. I actually ran tests, and it was working (forget about the detail that used the same regex both time and it wasn't going to match with old implementations).

If you have a better way to write the regex, please provide it.

**Comment by ivanpauno** [comment]
I closed and reopened to trigger the new checks.

**Comment by ivanpauno** [comment]
@rotu the check called `ubuntu-18,04, eloquent` seems to be using `master` and not `eloquent`.

Also, it's impossible that the checks using master would ever pass, if it's not possible to specify a branch in other package where we're doing simultaneous changes (e.g. `rmw`).
That's because we bump the API version is bumped once after the first API change, but not again after next release.

I'm not completely sure how to solve the problem, but I think we should only have actions for checking compatibility with past releases, and just rely on CI for master (which should be run manually by the user when the PR is ready).

**Comment by ivanpauno** [comment]
>@rotu the check called ubuntu-18,04, eloquent seems to be using master and not eloquent.

Sounds like https://github.com/ros2/rmw_cyclonedds/pull/133 fixed this.
But checks checking `source` (which I understand is `master`), will continue failing.

**Comment by rotu** [comment]
@ivanpauno I'm aware of the issue and it's due to my mistake in setting this up. Should be fixed by #133.

Yes, the checks ONLY check changes in this repo - the goal is to prevent accidental breakage. A change like this PR, involving a constellation of different commits, does tend to break things and presents a versioning challenge. Assuming that it would be too burdensome to preserve compatibility, the answer is to merge it with these checks failing.

**Comment by rotu** [comment]
FYI: here's me blundering through setting up CI. Using "source" was a mistake. https://github.com/ros-tooling/action-ros-ci/issues/120

**Comment by ivanpauno** [comment]
> Yes, the checks ONLY check changes in this repo - the goal is to prevent accidental breakage. A change like this PR, involving a constellation of different commits, does tend to break things and presents a versioning challenge. Assuming that it would be too burdensome to preserve compatibility, the answer is to merge it with these checks failing.

That sounds good to me.

I know that the actions using`source` will fail in this PR, but I would like to see if I'm maintaining compatibility with `dashing` and `eloquent` correctly.
Of course, I can manually run CI for `dashing` and `eloquent`, but that's what usually people don't do and end up breaking them.

Most of the PRs that have broken `dashing` and `eloquent` builds were merged with simultaneous changes in `rmw`, so ignoring the checks in those cases is ignoring the cases where we are most likely to break previous releases :smile: .

To avoid ignoring the result of the checks in those cases, I think we should either:
- Continue running `eloquent` and `dashing` checks if `source` check fails (i.e. not cancelling them).
- Delete the `source` checks, and rely on manually ran CI for `master`.

**Comment by rotu** [comment]
>     * Continue running `eloquent` and `dashing` checks if `source` check fails (i.e. not cancelling them).

Agreed. That *should* be the behavior since #133 (the setting `fail-fast: false` in the yml)

> Delete the source checks, and rely on manually ran CI for master.

Manually run CI is still part of the picture. The actions in this repo only check things *up to* `rmw_cyclonedds`. It would certainly be nice to test things upstream as well with GitHub Actions and add more tests to this repo, testing both correctness and performance.

**Comment by rotu** [comment]
To rerun this against latest CI workflow, I think you need to rebase onto master. I think if you fork this repo on Github, you can simply enable CI in the **Actions** tab of your fork and it will run on every push.

File: rmw_cyclonedds_cpp/src/rmw_node.cpp [hunk 9]
```diff
@@ -636,23 +672,36 @@ static void node_gone_from_domain_locked(dds_domainid_t did)
...
+  const char * namespace_, size_t domain_id
+#if !RMW_VERSION_GTE(0, 8, 2)
+  , const rmw_node_security_options_t * security_options
+#endif
 #if RMW_VERSION_GTE(0, 8, 1)
   , bool localhost_only
 #endif
 )
 {
-  static_cast<void>(context);
```
Comment (hidmic) [inline rmw_cyclonedds_cpp/src/rmw_node.cpp:655]: @ivanpauno isn't this up to the RMW version ?

File: rmw_cyclonedds_cpp/src/rmw_node.cpp (hunk 4 above)
Comment (hidmic) [reply rmw_cyclonedds_cpp/src/rmw_node.cpp:3088]: I think @wjwwood still applies.

**Comment by ivanpauno** [comment]
Note for reviewers: All checks are passing, except the ones using `master`. That's because there are simultaneous changes in others repos, and it's expected.

File: rmw_cyclonedds_cpp/src/rmw_node.cpp (hunk 4 above)
Comment (ivanpauno) [reply rmw_cyclonedds_cpp/src/rmw_node.cpp:3088]: The code is pre-existent, and I don't want to retrigger CI because of this. If you don't mind, I will ignore the comment.

File: rmw_cyclonedds_cpp/src/rmw_node.cpp (hunk 9 above)
Comment (ivanpauno) [reply rmw_cyclonedds_cpp/src/rmw_node.cpp:655]: A warning didn't appear in the `Dashing` and `Eloquent` checks, but you're right.
Fixed in https://github.com/ros2/rmw_cyclonedds/pull/106/commits/d213440d429a2638de95ead7f2dee1bdc03f229c.

File: rmw_cyclonedds_cpp/src/rmw_node.cpp (hunk 8 above)
Comment (ivanpauno) [reply rmw_cyclonedds_cpp/src/rmw_node.cpp]: Ok, finally fixed in https://github.com/ros2/rmw_cyclonedds/pull/106/commits/d213440d429a2638de95ead7f2dee1bdc03f229c.
I will rely on the `Eloquent` and `Dashing` checks here for validity, because the same code is exercised there.

**Comment by eboasson** [comment]
@ivanpauno I've finally carved out the time to work on this, but before really commencing I wanted to make sure there won't be any duplication of work, which there might be if you happen to also be looking into it. (If you are, I'm totally happy to leave you to it, by the way 😁)

Also looking at the scope of the changes, and as I have mentioned before, I think this is the point where we should stop providing compatibility with dashing and eloquent in rmw_cyclonedds master. Hopefully we'll be able to merge #123 into master first: that would then be a single version that supports all features and works with Dashing, Eloquent and Foxy, and so would also be valuable for anyone who needs to integrate multiple versions of ROS2 in a single system.

It'll take me some time to get all the changes done and tested locally, so I'd say there's plenty of time for updating #123 to match the current state of master and merging it.

**Comment by ivanpauno** [comment]
> @ivanpauno I've finally carved out the time to work on this, but before really commencing I wanted to make sure there won't be any duplication of work, which there might be if you happen to also be looking into it. (If you are, I'm totally happy to leave you to it, by the way grin)

Sounds good to me.
I'm not currently working on migrating the remaining rmw implementations to use one `Participant` per `Context`, neither I'm allocated to do it in a near future.

> Also looking at the scope of the changes, and as I have mentioned before, I think this is the point where we should stop providing compatibility with dashing and eloquent in rmw_cyclonedds master.

100% agreed, creating a new branch makes sense.

> Hopefully we'll be able to merge #123 into master first: that would then be a single version that supports all features and works with Dashing, Eloquent and Foxy, and so would also be valuable for anyone who needs to integrate multiple versions of ROS2 in a single system.

Sounds great to me!

----

If you need any reviews or if you have any questions, let me know, and I will be willing to help.
I will update https://github.com/ros2/design/pull/250 soon, so it better reflects what was done in `rmw_fastrtps`.
